libpod/lock/shm/shm_lock.c

```diff
@@ -0,0 +1,383 @@
+	  // Compute the number of the semaphore we are allocating
+	  sem_number = (BITMAP_SIZE * i) + num_within_bitmap;
+	  // OR in the bitmap
+	  shm->locks[i].bitmap = shm->locks[i].bitmap | test_map;
+	  // Clear the semaphore
+	  sem_post(&(shm->segment_lock));
+	  // Return the semaphore we've allocated
+	  return sem_number;
+	}
+	test_map = test_map << 1;
+	num_within_bitmap++;
+      }
+      // We should never fall through this loop
+      // TODO maybe an assert() here to panic if we do?
+    }
+  }
+
+  // Post to the semaphore to clear the lock
+  sem_post(&(shm->segment_lock));
+
+  // All bitmaps are full
+  // We have no available semaphores, report allocation failure
+  return -1 * ENOSPC;
+}
+
+// Deallocate a given semaphore
+// Returns 0 on success, negative ERRNO values on failure
+int32_t deallocate_semaphore(shm_struct_t *shm, uint32_t sem_index) {
+  bitmap_t test_map;
+  int bitmap_index, index_in_bitmap, ret_code, i;
+
+  if (shm == NULL) {
+    return -1 * EINVAL;
+  }
+
+  // Check if the lock index is valid
+  if (sem_index >= shm->num_locks) {
+    return -1 * EINVAL;
+  }
+
+  bitmap_index = sem_index / BITMAP_SIZE;
+  index_in_bitmap = sem_index % BITMAP_SIZE;
+
+  // This should never happen if the sem_index test above succeeded, but better
+  // safe than sorry
+  if (bitmap_index >= shm->num_bitmaps) {
+    return -1 * EFAULT;
+  }
+
+  test_map = 0x1;
+  for (i = 0; i < index_in_bitmap; i++) {
+    test_map = test_map << 1;
+  }
+
+  // Lock the semaphore controlling access to our shared memory
+  do {
+    ret_code = sem_wait(&(shm->segment_lock));
+  } while(ret_code == EINTR);
+  if (ret_code != 0) {
+    return -1 * errno;
+  }
+
+  // Check if the semaphore is allocated
+  if ((test_map & shm->locks[bitmap_index].bitmap) == 0) {
+    // Post to the semaphore to clear the lock
+    sem_post(&(shm->segment_lock));
+
+    return -1 * ENOENT;
+  }
+
+  // The semaphore is allocated, clear it
+  // Invert the bitmask we used to test to clear the bit
+  test_map = ~test_map;
+  shm->locks[bitmap_index].bitmap = shm->locks[bitmap_index].bitmap & test_map;
+
+  // Post to the semaphore to clear the lock
+  sem_post(&(shm->segment_lock));
+
+  return 0;
+}
+
+// Lock a given semaphore
+// Does not check if the semaphore is allocated - this ensures that, even for
+// removed containers, we can still successfully lock to check status (and
+// subsequently realize they have been removed).
+// Returns 0 on success, -1 on failure
+int32_t lock_semaphore(shm_struct_t *shm, uint32_t sem_index) {
+  int bitmap_index, index_in_bitmap, ret_code;
+
+  if (shm == NULL) {
+    return -1 * EINVAL;
+  }
+
+  if (sem_index >= shm->num_locks) {
+    return -1 * EINVAL;
+  }
+
+  bitmap_index = sem_index / BITMAP_SIZE;
+  index_in_bitmap = sem_index % BITMAP_SIZE;
+
+  // Lock the semaphore controlling access to our shared memory
+  do {
+    ret_code = sem_wait(&(shm->locks[bitmap_index].locks[index_in_bitmap]));
```

Comment:
If goroutine A is scheduled on thread T, locks a specific lock, and then goroutine B is scheduled to thread T and blocks here, is goroutine A guaranteed to be scheduled on a separate thread and eventually have the opportunity to unlock the lock?

Reply:
I don't think we can assume that. Also, the behavior of how goroutines are scheduled is an implementation detail :/

Reply:
POSIX mutexes have optional scheduling assistance to help prevent deadlocks with OS threads, but that doesn't really solve goroutine scheduling

Reply:
Per https://stackoverflow.com/questions/28354141/c-code-and-goroutine-scheduling CGo code does not block other goroutines from running - so it sounds like this shouldn't be an issue, as Go would allow the unlock goroutine to start.

Reply:
ACK; that link suggest that Cgo is treated like any possibly blocking syscall, and Go definitely must have _some_ mechanism to support an ~arbitrary number of concurrent coroutines which are blocked in a syscall for a very long time without blocking other CPU-heavy coroutines.